diff --git a/dwl.c b/dwl.c
index dc0437e..e0e9362 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1961,7 +1961,7 @@ pointerfocus(Client *c, struct wlr_surface *surface, double sx, double sy,
 {
 	struct timespec now;
 
-	if (surface != seat->pointer_state.focused_surface &&
+	if ((!active_constraint || active_constraint->surface != surface) &&
 			sloppyfocus && time && c && !client_is_unmanaged(c))
 		focusclient(c, 0);
 

