diff --git a/config.def.h b/config.def.h
index db0babc..8514d7e 100644
--- a/config.def.h
+++ b/config.def.h
@@ -19,6 +19,11 @@ static const float fullscreen_bg[]         = {0.1, 0.1, 0.1, 1.0}; /* You can al
 /* logging */
 static int log_level = WLR_ERROR;
 
+static const Env envs[] = {
+    /* variable            value */
+    { "XDG_CURRENT_DESKTOP",    "wlroots" },
+};
+
 static const Rule rules[] = {
 	/* app_id     title       tags mask     isfloating   monitor */
 	/* examples:
diff --git a/dwl.c b/dwl.c
index ef27a1d..de64d00 100644
--- a/dwl.c
+++ b/dwl.c
@@ -217,6 +217,11 @@ typedef struct {
 	int monitor;
 } Rule;
 
+typedef struct {
+    const char *variable;
+    const char *value;
+} Env;
+
 typedef struct {
 	struct wlr_scene_tree *scene;
 
@@ -1966,6 +1971,8 @@ run(char *startup_cmd)
 	if (!socket)
 		die("startup: display_add_socket_auto");
 	setenv("WAYLAND_DISPLAY", socket, 1);
+	for (int i = 0; i < LENGTH(envs); i++)
+			setenv(envs[i].variable, envs[i].value, 1);
 
 	/* Start the backend. This will enumerate outputs and inputs, become the DRM
 	 * master, etc */
