diff --git a/dwl.c b/dwl.c
index ef27a1d..58f6c73 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1248,12 +1248,14 @@ focusclient(Client *c, int lift)
 void
 focusmon(const Arg *arg)
 {
+	Monitor *prevm = selmon;
 	int i = 0, nmons = wl_list_length(&mons);
 	if (nmons)
 		do /* don't switch to disabled mons */
 			selmon = dirtomon(arg->i);
 		while (!selmon->wlr_output->enabled && i++ < nmons);
 	focusclient(focustop(selmon), 1);
+	wlr_cursor_move(cursor, NULL, selmon->m.x - prevm->m.x , 0);
 }
 
 void

