From c15fbbf1bfebc5c9311eeb893326f8269224b60b Mon Sep 17 00:00:00 2001
From: Palanix <palanixyt@gmail.com>
Date: Fri, 28 Jul 2023 02:14:32 +0200
Subject: [PATCH] Implement zoomswap

---
 dwl.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/dwl.c b/dwl.c
index 93f66efe..693a8b17 100644
--- a/dwl.c
+++ b/dwl.c
@@ -191,6 +191,7 @@ struct Monitor {
 	double mfact;
 	int nmaster;
 	char ltsymbol[16];
+	Client *prevc;
 };
 
 typedef struct {
@@ -907,6 +908,7 @@ createmon(struct wl_listener *listener, void *data)
 	size_t i;
 	Monitor *m = wlr_output->data = ecalloc(1, sizeof(*m));
 	m->wlr_output = wlr_output;
+	m->prevc = NULL;
 
 	wlr_output_init_render(wlr_output, alloc, drw);
 
@@ -2658,7 +2660,7 @@ xytonode(double x, double y, struct wlr_surface **psurface,
 void
 zoom(const Arg *arg)
 {
-	Client *c, *sel = focustop(selmon);
+	Client *c, *sel = focustop(selmon), *tmp = sel;
 
 	if (!sel || !selmon || !selmon->lt[selmon->sellt]->arrange || sel->isfloating)
 		return;
@@ -2679,9 +2681,12 @@ zoom(const Arg *arg)
 	/* If we passed sel, move c to the front; otherwise, move sel to the
 	 * front */
 	if (!sel)
-		sel = c;
+		sel = selmon->prevc ? selmon->prevc : c, c = tmp;
+	wl_list_remove(&c->link);
+	wl_list_insert(&sel->link, &c->link);
 	wl_list_remove(&sel->link);
 	wl_list_insert(&clients, &sel->link);
+	selmon->prevc = c;
 
 	focusclient(sel, 1);
 	arrange(selmon);
