From ef6f9aedba7316afa019e4314cfb41eceaa759f7 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Tue, 28 Nov 2023 11:31:50 -0500
Subject: [PATCH 1/2] apply main...guidocella:shiftview.patch

---
 config.def.h |  4 ++++
 shiftview.c  | 35 +++++++++++++++++++++++++++++++++++
 2 files changed, 39 insertions(+)
 create mode 100644 shiftview.c

diff --git a/config.def.h b/config.def.h
index 4e30885b7..b3c79b295 100644
--- a/config.def.h
+++ b/config.def.h
@@ -115,6 +115,8 @@ static const enum libinput_config_tap_button_map button_map = LIBINPUT_CONFIG_TA
 static const char *termcmd[] = { "foot", NULL };
 static const char *menucmd[] = { "bemenu-run", NULL };
 
+#include "shiftview.c"
+
 static const Key keys[] = {
 	/* Note that Shift changes certain key codes: c -> C, 2 -> at, etc. */
 	/* modifier                  key                 function        argument */
@@ -128,6 +130,8 @@ static const Key keys[] = {
 	{ MODKEY,                    XKB_KEY_l,          setmfact,       {.f = +0.05} },
 	{ MODKEY,                    XKB_KEY_Return,     zoom,           {0} },
 	{ MODKEY,                    XKB_KEY_Tab,        view,           {0} },
+	{ MODKEY,                    XKB_KEY_a,          shiftview,      { .i = -1 } },
+	{ MODKEY,                    XKB_KEY_semicolon,  shiftview,      { .i = 1 } },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_C,          killclient,     {0} },
 	{ MODKEY,                    XKB_KEY_t,          setlayout,      {.v = &layouts[0]} },
 	{ MODKEY,                    XKB_KEY_f,          setlayout,      {.v = &layouts[1]} },
diff --git a/shiftview.c b/shiftview.c
new file mode 100644
index 000000000..6903a5c37
--- /dev/null
+++ b/shiftview.c
@@ -0,0 +1,35 @@
+// "arg->i" stores the number of tags to shift right (positive value)
+//          or left (negative value)
+void
+shiftview(const Arg *arg)
+{
+	Arg a;
+	Client *c;
+	size_t ntags = tagcount;
+	bool visible = false;
+	int i = arg->i;
+	int count = 0;
+	int nextseltags, curseltags = selmon->tagset[selmon->seltags];
+
+	do {
+		if (i > 0) // left circular shift
+			nextseltags = (curseltags << i) | (curseltags >> (ntags - i));
+		else // right circular shift
+			nextseltags = curseltags >> (- i) | (curseltags << (ntags + i));
+
+		// Check if the tag is visible
+		wl_list_for_each(c, &clients, link) {
+			if (c->mon == selmon && nextseltags & c->tags) {
+				visible = true;
+				break;
+			}
+		}
+
+		i += arg->i;
+	} while (!visible && ++count <= ntags);
+
+	if (count <= ntags) {
+		a.i = nextseltags;
+		view(&a);
+	}
+}

From 41ba97a0b5081e3beb7e1d2eaa46b7f62e47c60d Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Tue, 28 Nov 2023 11:32:23 -0500
Subject: [PATCH 2/2] fix patch

replace tagcount with TAGCOUNT
---
 shiftview.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/shiftview.c b/shiftview.c
index 6903a5c37..dfb7c5e30 100644
--- a/shiftview.c
+++ b/shiftview.c
@@ -5,7 +5,7 @@ shiftview(const Arg *arg)
 {
 	Arg a;
 	Client *c;
-	size_t ntags = tagcount;
+	size_t ntags = TAGCOUNT;
 	bool visible = false;
 	int i = arg->i;
 	int count = 0;
