From 49a2276f7fd348b68192494e53d700f759a0a96e Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Fri, 5 Jul 2024 11:47:01 -0500
Subject: [PATCH] implement env

---
 config.def.h | 5 +++++
 dwl.c        | 7 +++++++
 2 files changed, 12 insertions(+)

diff --git a/config.def.h b/config.def.h
index 22d2171..acc7882 100644
--- a/config.def.h
+++ b/config.def.h
@@ -20,6 +20,11 @@ static const float fullscreen_bg[]         = {0.1f, 0.1f, 0.1f, 1.0f}; /* You ca
 /* logging */
 static int log_level = WLR_ERROR;
 
+static const Env envs[] = {
+    /* variable            value */
+    { "XDG_CURRENT_DESKTOP",    "wlroots" },
+};
+
 /* NOTE: ALWAYS keep a rule declared even if you don't use rules (e.g leave at least one example) */
 static const Rule rules[] = {
 	/* app_id             title       tags mask     isfloating   monitor */
diff --git a/dwl.c b/dwl.c
index dc0437e..0490d44 100644
--- a/dwl.c
+++ b/dwl.c
@@ -232,6 +232,11 @@ typedef struct {
 	int monitor;
 } Rule;
 
+typedef struct {
+    const char *variable;
+    const char *value;
+} Env;
+
 typedef struct {
 	struct wlr_scene_tree *scene;
 
@@ -2168,6 +2173,8 @@ run(char *startup_cmd)
 	if (!socket)
 		die("startup: display_add_socket_auto");
 	setenv("WAYLAND_DISPLAY", socket, 1);
+	for (int i = 0; i < (int)LENGTH(envs); i++)
+			setenv(envs[i].variable, envs[i].value, 1);
 
 	/* Start the backend. This will enumerate outputs and inputs, become the DRM
 	 * master, etc */
-- 
2.45.1

